### PR TITLE
Implement Kubernetes Readiness probe

### LINKED ISSUE #4306: Provide Kubernetes probes
# What should be done?

What documentation in #4424 says.

# DoD

- Implement probes as documentation says.

### PR DESCRIPTION
**Pull request by noisersup**
# Description

**PR not cleared and finished, it's a PoC of the probe to discuss the approach.**

Closes #4306.

## Readiness checklist

- [ ] I added/updated unit tests (and they pass).
- [ ] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [ ] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [ ] I ensured that PR title is good enough for the changelog.
- [ ] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [ ] I marked all done items in this checklist.

### CODE REVIEW

File: internal/util/debug/debug.go [hunk 1]
```diff
@@ -118,8 +128,22 @@ func Listen(opts *ListenOpts) (*Handler, error) {
...
+
+		ctx := conninfo.Ctx(r.Context(), connInfo)
+
+		_, err := opts.Backend.Status(ctx, nil)
+		if err != nil {
+			opts.L.Error("Backend returned error", zap.Error(err))
+			rw.WriteHeader(http.StatusServiceUnavailable)
+
+			return
+		}
```
Comment (noisersup) [inline internal/util/debug/debug.go:145]: The backend is only provided if new authentication is enabled. If the old one is used, the check just returns straightforward `200`.

Providing backend, allows us to easily call `Status` which properly pings the database.

File: cmd/ferretdb/main.go [hunk 2]
```diff
@@ -519,6 +498,33 @@ func run() {
...
+
+			h, err := debug.Listen(opts)
+			if err != nil {
+				l.Sugar().Fatalf("Failed to create debug handler: %s.", err)
+			}
+
+			h.Serve(ctx)
+		}()
+	}
+
```
Comment (noisersup) [inline cmd/ferretdb/main.go:527]: It's possible to not move the whole debug section from the beginning, by calling `h.Serve()`, `l.Run()`, and `r.Run()`, later in the code. If the approach is alright, I'll proceed to do that, for now less changes is better.

File: internal/util/debug/debug.go (hunk 1 above)
Comment (chilagrow) [reply internal/util/debug/debug.go:145]: Yeah for backend authentication, it won't be able to ping.

File: cmd/ferretdb/main.go (hunk 2 above)
Comment (chilagrow) [reply cmd/ferretdb/main.go:527]: What is the issue here? Moving the code? :thinking:

File: cmd/ferretdb/main.go [hunk 3]
```diff
@@ -519,6 +498,33 @@ func run() {
...
+			l := logger.Named("debug")
+
+			opts := &debug.ListenOpts{
+				TCPAddr: cli.DebugAddr,
+				L:       l,
+				R:       metricsRegisterer,
+			}
+
+			if cli.Setup.Database != "" {
```
Comment (chilagrow) [inline cmd/ferretdb/main.go:515]: `cli.Setup` may be nil

Comment (AlekSi) [reply cmd/ferretdb/main.go:515]: Nope, that's struct value, not pointer

File: cmd/ferretdb/main.go (hunk 2 above)
Comment (noisersup) [reply cmd/ferretdb/main.go:527]: I believe we'd like to make sure that the debug handler is the first that is running

File: internal/util/debug/debug.go [hunk 4]
```diff
@@ -125,8 +128,28 @@ func Listen(opts *ListenOpts) (*Handler, error) {
 		rw.WriteHeader(http.StatusOK)
 	})
 
-	readyHandler := http.HandlerFunc(func(rw http.ResponseWriter, _ *http.Request) {
-		// TODO https://github.com/FerretDB/FerretDB/issues/4306
+	// ready handler, which is used for readiness probe,
+	// returns StatusOK when Backend.Status() returned no error.
```
Comment (chilagrow) [inline internal/util/debug/debug.go:132]: First two comment seems redundant of the implementation.

**Review by chilagrow — COMMENTED** [review]
Just small comment

**Pull request closed without merging**